compiler/rustc_builtin_macros/src/source_util.rs

```diff
@@ -215,25 +206,57 @@ pub fn expand_include_bytes(
     tts: TokenStream,
 ) -> Box<dyn MacResult + 'static> {
     let sp = cx.with_def_site_ctxt(sp);
-    let file = match get_single_str_from_tts(cx, sp, tts, "include_bytes!") {
-        Ok(file) => file,
+    let (path, path_span) = match get_single_str_spanned_from_tts(cx, sp, tts, "include_bytes!") {
+        Ok(res) => res,
         Err(guar) => return DummyResult::any(sp, guar),
     };
-    let file = match resolve_path(&cx.sess, file.as_str(), sp) {
-        Ok(f) => f,
-        Err(err) => {
-            let guar = err.emit();
-            return DummyResult::any(sp, guar);
-        }
-    };
-    match cx.source_map().load_binary_file(&file) {
+    match load_binary_file(cx, path.as_str(), sp, path_span) {
         Ok(bytes) => {
             let expr = cx.expr(sp, ast::ExprKind::IncludedBytes(bytes));
             MacEager::expr(expr)
         }
-        Err(e) => {
-            let guar = cx.dcx().span_err(sp, format!("couldn't read {}: {}", file.display(), e));
-            DummyResult::any(sp, guar)
+        Err(dummy) => dummy,
+    }
+}
+
+fn load_binary_file(
+    cx: &mut ExtCtxt<'_>,
+    original_path: &str,
+    macro_span: Span,
+    path_span: Span,
+) -> Result<Lrc<[u8]>, Box<dyn MacResult>> {
+    let resolved_path = match resolve_path(&cx.sess, original_path, macro_span) {
+        Ok(path) => path,
+        Err(err) => {
+            let guar = err.emit();
+            return Err(DummyResult::any(macro_span, guar));
+        }
+    };
+    match cx.source_map().load_binary_file(&resolved_path) {
+        Ok(data) => Ok(data),
+        Err(io_err) => {
+            let mut err = cx.dcx().struct_span_err(
+                macro_span,
+                format!("couldn't read {}: {io_err}", resolved_path.display()),
+            );
+            if Path::new(original_path).is_relative() {
+                for prefix in ["..", "../.."] {
```

Comment:
instead of this very limited check, would it make more sense to canonicslize the path and then walk the parents?
It would then have to compute the amount of .. which is also not very nice but at least it can do more than this. We can't walk forever as /.. is / on Unix systems but maybe what about just having a loop that prepends .. like 10 times? Don't know which solution is the nicest but I think we can do a lot better than just trying 2.
I don't think there's any security concern with probing as much as we can.

Reply:
Stripping the path of all the `..` components and then testing all the paths up to `/` by counting the number of components in the canonicalize of the CWD should give us a pretty solid approximation.

See this [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=253a43b9c3ee41650fda8357e00da0f5) for a example.

Reply:
[`std::path::absolute`](https://doc.rust-lang.org/std/path/fn.absolute.html)?

Reply:
In my example, the canonicalize is done on the current working directory, and is only here to have an upper bound on the number of ".." to add. That upper bound may indeed be wrong with symlinks but it's not necessary to be 100% right.